acyclic/test/src-3/acyclic/TestUtils.scala

```diff
@@ -0,0 +1,122 @@
+import dotty.tools.dotc.config.ScalaSettings
+import dotty.tools.dotc.core.Contexts.{Context, ContextBase, FreshContext, NoContext}
+import dotty.tools.dotc.interfaces.Diagnostic.{ERROR, INFO, WARNING}
+import dotty.tools.dotc.plugins.Plugin
+import dotty.tools.dotc.reporting.{ConsoleReporter, StoreReporter}
+import java.net.URLClassLoader
+import java.nio.file.Paths
+import utest._
+import utest.asserts._
+
+object TestUtils extends BaseTestUtils {
+  val srcDirName: String = "src-3"
+
+  /**
+   * Attempts to compile a resource folder as a compilation run, in order
+   * to test whether it succeeds or fails correctly.
+   */
+  def make(
+      path: String,
+      extraIncludes: Seq[String] = Seq("acyclic/src/acyclic/package.scala"),
+      force: Boolean = false,
+      warn: Boolean = false,
+      collectInfo: Boolean = true
+  ): Seq[(String, String)] = {
+    val src = "acyclic/test/resources/" + path
+    val sources = (getFilePaths(src) ++ extraIncludes).map(f => PlainFile(Path(Paths.get(f))))
+    val vd = new VirtualDirectory("(memory)", None)
+    val loader = getClass.getClassLoader.asInstanceOf[URLClassLoader]
+    val entries = loader.getURLs map (_.getPath)
+
+    val scalaSettings = new ScalaSettings {}
+    val settingsState1 = scalaSettings.outputDir.updateIn(scalaSettings.defaultState, vd)
+    val settingsState2 = scalaSettings.classpath.updateIn(settingsState1, ClassPath.join(entries*))
+
+    val opts = List(
+      if (force) Seq("force") else Seq(),
+      if (warn) Seq("warn") else Seq()
+    ).flatten
+
+    val settingsState3 = if (opts.nonEmpty) {
+      val options = opts.map("acyclic:" + _)
+      println("options: " + options)
+      scalaSettings.pluginOptions.updateIn(settingsState2, options)
+    } else {
+      settingsState2
+    }
+
+    var cycles: Option[Seq[Seq[(Value, SortedSet[Int])]]] = None
+    val storeReporter = if (collectInfo) Some(new StoreReporter()) else None
+
+    val ctxBase = new ContextBase {
+      override val initialCtx: Context = FreshContext.initial(NoContext.base, settings)
+
+      override protected def loadRoughPluginsList(using Context): List[Plugin] =
+        List(new plugin.TestPlugin(foundCycles =>
+          cycles = cycles match {
+            case None => Some(Seq(foundCycles))
+            case Some(oldCycles) => Some(oldCycles :+ foundCycles)
+          }
+        ))
+    }
+
+    given ctx: Context = FreshContext.initial(ctxBase, new ScalaSettings {
+      override val defaultState = settingsState3
+    })
+      .asInstanceOf[FreshContext]
+      .setReporter(storeReporter.getOrElse(ConsoleReporter()))
+
+    ctx.initialize()
+
+    val compiler = new Compiler()
+    val run = compiler.newRun
+
+    run.compile(sources)
+
+    if (vd.toList.isEmpty) throw CompilationException(cycles.get)
+
+    storeReporter.map(_.pendingMessages.toSeq.map(i => (i.msg.message, i.level match {
+      case ERROR => "error"
+      case INFO => "info"
+      case WARNING => "warning"
```

Comment:
Scala 3 diagnostics don't have a `severity`, instead they have a `level: Int` that maps to one of these 3